src/docs/arch-hs.dox

```diff
@@ -4,5 +4,111 @@ The History Store in WiredTiger tracks old (all but the latest committed)
+filename for the History Stable table.
+
+@section arch_hs_cursor Historty Store cursor interface
+WiredTiger uses a cursor interface for History Store table to make it easier for different modules
+to perform data operations on the History Strore table. A new cursor can be opened by calling
+\c __wt_curhs_open(). History Store cursor implementation supports most of the standard \c WT_CURSOR
+methods, with one key exception being \c WT_CURSOR::search() method. Recall that the key contains a
+counter value and it is expected that caller can not possibly know the correct counter value to use.
+Therefore, \c WT_CURSOR::search_near() is used for all search operations. Two helper functions,
+\c __wt_curhs_search_near_before() and \c __wt_curhs_search_near_after(), have been provided to
+facilitate searching the required record in the History Store table.
+
+@section arch_hs_cursor_visibilty Historty Store cursor interface and visibility rules
+When using the Historty Store cursor interface, the user can configure the type of visibility checks
+that are performed on the records. The behavhiour is controlled by a set of cursor flags:
+
+- When the flag \c WT_CURSTD_HS_READ_ALL is set on the cursor, no visibility checks are performed
+on the records returned to the API user. This means that cursor interface can even return a record
+with a globally visible tombstone. When this flag is set, it suppresses the effect of other
+visibility flags.
+
+- When the flag \c WT_CURSTD_HS_READ_COMMITTED is set, cursor interface can return any record except
+for records that have a globally visible tombstone. This flag has no effect when the flag
+\c WT_CURSTD_HS_READ_ALL is set.
+
+- When none of the flags discussed above are set on the cursor, snapshot based visibility checks are
+performed on all records returned to the API user. History Store cursor users must use one of
+the flags discussed above if the caller thread is running in a lower isolation level and
+doesn't hold a valid snapshot.
+
+@section arch_hs_reconciliation Historty Store and reconciliation
+When a modified user page is processed by either checkpoint or eviction module, the reconciliation
+code examines the update chain for each key and decides on the latest committed update that will be
+written as the on-disk value for the key. All the updates that are older than the on-disk version
+but not yet globally visible are added to the History Store table. The reconciliation code also
+fixes out-of-order timestamps for updates being added to the History Store table.
+
+When a modified History Store table page is processed by reconciliation, certain records on the page
+are removed when creating a new page image if the tombstones on those records have become gobally
+visible. This ensures that WiredTiger is only keeping the relevant historical data required to
+serve the oldest reader in the system or as dictated by the oldest timestamp set by the application.
```

Comment:
Probably worth mentioning that once all records on the page are obsolete the page itself is removed, allowing the history store to not grow too large. 

Side note: A reference to the oldest timestamp may exist in other  architecture guide pages and could be referenced to describe what makes something globally obsolete.

Reply:
I have added a reference to another section that describes timestamps.